crates/core/action-pipeline/src/actions/run_task.rs

```diff
@@ -94,18 +94,34 @@ pub async fn run_task(
             .insert(target.clone(), TargetState::Passthrough);
     }
 
-    let attempts_result = if is_cache_enabled {
-        let context = context.read().await;
-
-        runner.create_and_run_command(&context, runtime).await
-    } else {
-        // Concurrent long-running tasks will cause a deadlock, as some threads will
-        // attempt to write to context while others are reading from it, and long-running
-        // tasks may never release the lock. Unfortuantely we have to clone here to work
-        // around it, so revisit in the future.
-        let context = (context.read().await).clone();
-
-        runner.create_and_run_command(&context, runtime).await
+    let attempts_result = {
+        let _ctx: Arc<RwLock<ActionContext>>;
+        let ctx = if is_cache_enabled {
+            &context
+        } else {
+            // Concurrent long-running tasks will cause a deadlock, as some threads will
+            // attempt to write to context while others are reading from it, and long-running
+            // tasks may never release the lock. Unfortuantely we have to clone here to work
+            // around it, so revisit in the future.
+            _ctx = Arc::new(RwLock::new(context.read().await.clone()));
```

Comment:
This is the part that I don't like.

Although creating an `Arc<RwLock<*>>` instance does help to match the types of both branches, it involves 2 extra allocations, and 2 extra indirections once we access the value a few lines below.